File: src/pipeline/orchestrator.ts
import type { GenerateResponse, Pass, ProgressEvent } from "../api/contract";
import type { SceneState } from "../scene/schema";
import { layout } from "./layout";
import { geometryCheck } from "./geometryCheck";
import { fix } from "./fix";
import { claudePlanner } from "./claudePlanner";
import { meshyAssetProvider } from "./meshyAssetProvider";
import { claudeVisionCritic } from "./claudeVisionCritic";
import { launchBrowser, type RenderInput } from "../render/headless";
import { captureViews } from "../render/multiangle/index";
import { criticCameras } from "../render/criticCameras";
import { deriveResponseKey, readCachedResponse, writeCachedResponse } from "./responseCache";
import { getOrCreatePlan } from "./planCache";

// The agentic loop: plan -> assets -> layout -> [render -> geometry + vision ->
// fix] x amendRounds. The response always carries amendRounds + 1 passes (the
// draft plus one per amend round); every pushed sceneState is a deep clone so the
// mutable working object is never shared across passes.

const ASSET_CONCURRENCY = 3;

// Cache-aware entry point. An identical request — same prompt + amendRounds — is
// served from the on-disk response cache: the first run is live (minutes), every
// repeat is sub-second and spends ZERO Meshy credits. DREAM3D_RESPONSE_CACHE=0
// bypasses the cache entirely (both read and write). `onEvent` streams progress
// on a live run; a cache HIT returns instantly and emits no events — the server
// completes the job on promise resolution, not on a terminal event, so an empty
// progress log is correct.
export async function generate(
  prompt: string,
  amendRounds: number,
  onEvent?: (ev: ProgressEvent) => void,
): Promise<GenerateResponse> {
  const useCache = process.env.DREAM3D_RESPONSE_CACHE !== "0";
  if (useCache) {
    const key = deriveResponseKey(prompt, amendRounds);
    const cached = readCachedResponse(key);
    if (cached) {
      console.log(`[dream3d] response cache HIT ${key} (amendRounds=${amendRounds})`);
      return cached;
    }
    const result = await runPipeline(prompt, amendRounds, onEvent);
    writeCachedResponse(key, { prompt, amendRounds }, result);
    return result;
  }
  return runPipeline(prompt, amendRounds, onEvent);
}

// The live pipeline. `onEvent` is threaded straight through so a live run still
// streams progress.
async function runPipeline(
  prompt: string,
  amendRounds: number,
  onEvent?: (ev: ProgressEvent) => void,
): Promise<GenerateResponse> {
  onEvent?.({ kind: "plan" });
  const plan = await getOrCreatePlan(prompt, () => claudePlanner.plan(prompt));
  onEvent?.({ kind: "plan_done", objectCount: plan.objects.length });

  const total = plan.objects.length;
  let completed = 0; // single-threaded JS: no race on the shared counter across concurrent assets.
  const assets = await mapWithConcurrency(plan.objects, ASSET_CONCURRENCY, async (obj, index) => {
    onEvent?.({ kind: "asset_start", index, total, label: obj.label });
    const asset = await meshyAssetProvider.generate(obj);
    completed++;
    onEvent?.({ kind: "asset_done", index, total, completed, label: obj.label });
    return asset;
  });

  let scene = layout(plan);
  onEvent?.({ kind: "layout" });
  scene.objects.forEach((obj, index) => {
    obj.glbUrl = assets[index].glbUrl;
    obj.status = "ready";
  });

  const passes: Pass[] = [{ sceneState: structuredClone(scene) }];

  // Render the working scene from several angles each round so the vision critic
  // judges a correctly-scaled, multi-view scene; one warm browser is reused across
  // every round and closed in the finally.
  const browser = await launchBrowser();
  try {
    for (let round = 1; round <= amendRounds; round++) {
      onEvent?.({ kind: "render", round });
      const views = await captureSceneViews(scene, browser);
      const issues = [...geometryCheck(scene), ...(await claudeVisionCritic.review({ scene, views }))];
      onEvent?.({ kind: "critique", round, issueCount: issues.length });
      if (issues.length === 0) {
        onEvent?.({ kind: "clean", round });
        break;
      }
      scene = fix(scene, issues);
      onEvent?.({ kind: "fix", round, issueCount: issues.length });
      passes.push({ sceneState: structuredClone(scene) });
    }
  } finally {
    await browser.close();
  }

  onEvent?.({ kind: "done", passCount: passes.length });
  return { passes };
}

// Render the working scene from the critic's framing angles over the warm
// browser, returning one labeled PNG data URL per view for the vision critic.
async function captureSceneViews(scene: SceneState, browser: any): Promise<{ name: string; dataUrl: string }[]> {
  const shots = await captureViews(toRenderInput(scene), criticCameras(scene.room), { browser });
  return shots.map((s) => ({ name: s.name, dataUrl: `data:image/png;base64,${s.png.toString("base64")}` }));
}

// Maps the working SceneState -> the headless render harness's RenderInput, carrying
// approxSize so the render page normalizes each model to its intended bbox.
function toRenderInput(scene: SceneState): RenderInput {
  return {
    room: scene.room,
    objects: scene.objects.map((obj) => ({
      glbUrl: obj.glbUrl,
      primitive: obj.glbUrl ? undefined : "box",
      position: obj.transform.position,
      rotationYDeg: obj.transform.rotationYDeg,
      scale: obj.transform.scale,
      approxSize: obj.approxSize,
    })),
  };
}

// Run `fn` over `items` with at most `limit` in flight; results keep input order.
async function mapWithConcurrency<T, R>(items: T[], limit: number, fn: (item: T, index: number) => Promise<R>): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let cursor = 0;
  async function worker(): Promise<void> {
    while (cursor < items.length) {
      const index = cursor++;
      results[index] = await fn(items[index], index);
    }
  }
  const pool = Array.from({ length: Math.min(limit, items.length) }, () => worker());
  await Promise.all(pool);
  return results;
}

File: backend/src/pipeline/orchestrator.py
"""Cache-aware Dream3D generation orchestrator."""

from __future__ import annotations

import asyncio
import base64
import os
from collections.abc import Callable
from typing import Any

from log.audit import current_run_id
from pipeline import asset_provider, claude_planner, claude_vision_critic
from pipeline.fix import fix
from pipeline.geometry_check import geometry_check
from pipeline.layout import layout
from pipeline.plan_cache import get_or_create_plan
from pipeline.response_cache import (
    derive_response_key,
    read_cached_response,
    write_cached_response,
)
from render.critic_cameras import critic_cameras
from render.headless import launch_browser
from render.multiangle import capture_views
from scene.schema import GenerateResponse, Pass, PlannedObject, SceneState

ASSET_CONCURRENCY = 20

ProgressCallback = Callable[[dict[str, Any]], None]


async def generate(
    prompt: str,
    amend_rounds: int,
    on_event: ProgressCallback | None = None,
) -> GenerateResponse:
    """Run the full generation pipeline, using the response cache when enabled."""
    use_cache = os.environ.get("DREAM3D_RESPONSE_CACHE") != "0"
    if use_cache:
        key = derive_response_key(prompt, amend_rounds)
        cached = read_cached_response(key)
        if cached is not None:
            print(f"[dream3d] response cache HIT {key} (amendRounds={amend_rounds})")
            _emit(on_event, {"kind": "cached"})
            return GenerateResponse.model_validate(cached.model_dump(by_alias=True))

        result = await _run_pipeline(prompt, amend_rounds, on_event)
        write_cached_response(
            key,
            prompt=prompt,
            amend_rounds=amend_rounds,
            response=result,
        )
        return result

    return await _run_pipeline(prompt, amend_rounds, on_event)


async def _run_pipeline(
    prompt: str,
    amend_rounds: int,
    on_event: ProgressCallback | None,
) -> GenerateResponse:
    _emit(on_event, {"kind": "plan"})
    plan = await get_or_create_plan(prompt, lambda: claude_planner.plan(prompt))
    _emit(on_event, {"kind": "plan_done", "object_count": len(plan.objects)})

    paths = await _generate_assets(plan.objects, on_event)

    scene = layout(plan)
    _emit(on_event, {"kind": "layout"})
    for index, obj in enumerate(scene.objects):
        obj.glb_url = paths[index]
        obj.status = "ready"

    passes = [Pass(scene_state=scene.model_copy(deep=True))]

    browser = await launch_browser()
    try:
        for round_ in range(1, amend_rounds + 1):
            _emit(on_event, {"kind": "render", "round": round_})
            views = await _capture_scene_views(scene, browser, on_event)
            issues = geometry_check(scene) + await claude_vision_critic.review(
                scene, views
            )
            _emit(
                on_event,
                {"kind": "critique", "round": round_, "issue_count": len(issues)},
            )
            if len(issues) == 0:
                _emit(on_event, {"kind": "clean", "round": round_})
                break
            scene = fix(scene, issues)
            _emit(
                on_event,
                {"kind": "fix", "round": round_, "issue_count": len(issues)},
            )
            passes.append(Pass(scene_state=scene.model_copy(deep=True)))
    finally:
        await browser.close()

    _emit(on_event, {"kind": "done", "pass_count": len(passes)})
    return GenerateResponse(passes=passes)


async def _generate_assets(
    objects: list[PlannedObject],
    on_event: ProgressCallback | None,
) -> list[str]:
    semaphore = asyncio.Semaphore(ASSET_CONCURRENCY)
    total = len(objects)
    completed = 0

    async def generate_one(index: int, obj: PlannedObject) -> str:
        nonlocal completed
        async with semaphore:
            _emit(
                on_event,
                {
                    "kind": "asset_start",
                    "index": index,
                    "total": total,
                    "label": obj.label,
                },
            )
            await asyncio.sleep(0)
            path = await asset_provider.generate(obj)
            completed += 1
            _emit(
                on_event,
                {
                    "kind": "asset_done",
                    "index": index,
                    "total": total,
                    "completed": completed,
                    "label": obj.label,
                },
            )
            return path

    return await asyncio.gather(
        *(generate_one(index, obj) for index, obj in enumerate(objects))
    )


async def _capture_scene_views(
    scene: SceneState,
    browser: Any,
    on_event: ProgressCallback | None,
) -> list[dict[str, str]]:
    shots = await capture_views(
        _to_render_input(scene),
        critic_cameras(scene.room),
        options={"browser": browser, "job_id": current_run_id()},
    )
    for shot in shots:
        if shot.blank_warning:
            _emit(
                on_event,
                {
                    "kind": "blank_warning",
                    "view": shot.name,
                    "warning": shot.blank_warning,
                },
            )
    return [
        {
            "name": shot.name,
            "dataUrl": "data:image/png;base64,"
            + base64.b64encode(shot.png).decode("ascii"),
        }
        for shot in shots
    ]


def _to_render_input(scene: SceneState) -> dict[str, Any]:
    return {
        "room": scene.room.model_dump(),
        "objects": [
            {
                "glbUrl": obj.glb_url,
                "primitive": "box" if not obj.glb_url else None,
                "position": obj.transform.position,
                "rotationYDeg": obj.transform.rotation_y_deg,
                "scale": obj.transform.scale,
                "approxSize": obj.approx_size,
            }
            for obj in scene.objects
        ],
    }


def _emit(on_event: ProgressCallback | None, event: dict[str, Any]) -> None:
    if on_event is not None:
        on_event(event)
